Add file containing release notes, that currently only include the re… (#1688)
dafny-lang/dafny: RELEASE_NOTES.md

# 3.4 

> The top four improvements in Dafny 3.4 are:

Reviewer: I don't think we need this line.

Author: I'm inclined to agree. I modelled this on the 3.3 release notes that do include a summary section, although they're significantly bigger in total: https://github.com/dafny-lang/dafny/releases/tag/v3.3.0

Still think we should leave it out?

Reviewer: I like the headline features, but I don't think they need an introduction.

Reviewer 2: I definitely like highlighting the features at least.  I don't have a strong opinion about this exact line.

- For certain classes of changes, prevent unexpected changes in verification behavior when changing a Dafny program 
- Add command line options to assist in debugging verification performance.
- Critical fixes to the IDE and greatly improved responsiveness of non-verification IDE features. 

> - The C# back-end supports traits as type parameters on datatypes.

Reviewer 3: Add a note that compiled downcasts are supported for these only for co-variant and non-variant type parameters, not yet contra-variant type parameters. This same note also applies under the `C#` section below (or perhaps this detail should only be mentioned in the `C#` section below, keeping this list of "top improvements" simpler to read).

### IDE 
- fix: Clean up process resources if IDE closed or restarted
- fix: Do not let the Dafny compilation status bar get in a stuck state

- feat: Verification status reporting shows which proof is being verified, which can help debug slow to verify proofs.
- feat: Publish parsing and resolution diagnostics before verification has completed. Verification diagnostics from previous runs are migrated.
- feat: Enable 'go to definition', 'hover' and 'signature help' features before verification has completed
- feat: Improve the hover feature to work for a wider scope of Dafny constructs, including function and method parameters, forall, exists and let expressions, and set and map comprehensions.
- feat: Add an experimental verification caching feature, which enables automatically determining which proofs need to verify again after making changes.

### UX
- feat: Improve error reporting when providing wrongly typed arguments in a function call
- feat: Improve error reporting when using type tests

### C#
- fix: resolve an instance where Dafny would produce invalid C# code: https://github.com/dafny-lang/dafny/issues/1607

> - feat: support traits as type parameters on datatypes for co-variant and non-variant type parameter, but not for contra-variant type parameters: https://github.com/dafny-lang/dafny/issues/1499

Reviewer: ```suggestion
- feat: support for variant type parameters on datatype definitions, with the omission of downcasting ones with conta-variant type parameters, enabling
traits as type arguments: https://github.com/dafny-lang/dafny/pull/1645
```
Please include the changes in #1645.

Author: I've left out the PR reference, since the rest of the notes don't reference PRs either. I think it makes more sense to reference issue than PRs, don't you? The issues can then reference the PRs again.

## Breaking changes

> - Proofs such as methods and lemmas whose verification behavior, either the result or the verification time, depend on arbitrary behavior in Dafny's solver, may show different verification behavior in Dafny 3.4. For customers who are not ready to change those proofs to make their verification more reliable, we offer the command line option `/mimicVerificationOf:3.3` to keep the Dafny 3.4 verification behavior consistent with 3.3.

Reviewer: Is _customers_ the language we want here?

- In Dafny 3.3, comprehensions quantified over subset types did not validate the constraint of the subset type, which could result in crashes at run-time. In 3.4, subset types are disabled in set comprehensions in compiled contexts, unless the subset constraint is itself compilable.

> Before, the following code would pass Dafny and be compiled without error, but would crash at run-time:

Reviewer 4: This part should be indented if you want it to be part of the bullet point above

```
type RefinedData = x: Data | ghostFunction(x)
method Main() {
  var s: set<Data> = ...
  var t = set x: RefinedData | x in s;
  forall x in t {
    if !ghostFunction(x) {
      var crash := 1/0;
    }
  }
}
```
In Dafny 3.4, the same code triggers a resolution error of the form:

> Error: RefinedData is a subset type and its constraint is not compilable, hence it cannot yet be used as the type of a bound variable in set comprehension. The next error will explain why the constraint is not compilable.

Reviewer 4: This could be a quote or a code block

# 3.4 

The top four improvements in Dafny 3.4 are:
- For certain classes of changes, prevent unexpected changes in verification behavior when changing a Dafny program 
- Add command line options to assist in debugging verification performance.
- Critical fixes to the IDE and greatly improved responsiveness of non-verification IDE features. 

> - The C# back-end partially supports traits as type parameters on datatypes.

Reviewer: ```suggestion
- The C# back-end supports traits as type parameters on datatypes.
```

### IDE 
- fix: Clean up process resources if IDE closed or restarted
- fix: Do not let the Dafny compilation status bar get in a stuck state

- feat: Verification status reporting shows which proof is being verified, which can help debug slow to verify proofs.
- feat: Publish parsing and resolution diagnostics before verification has completed. Verification diagnostics from previous runs are migrated.
- feat: Enable 'go to definition', 'hover' and 'signature help' features before verification has completed
- feat: Improve the hover feature to work for a wider scope of Dafny constructs, including function and method parameters, forall, exists and let expressions, and set and map comprehensions.
- feat: Add an experimental verification caching feature, which enables automatically determining which proofs need to verify again after making changes.

### UX
- feat: Improve error reporting when providing wrongly typed arguments in a function call
- feat: Improve error reporting when using type tests

### C#
- fix: resolve an instance where Dafny would produce invalid C# code. https://github.com/dafny-lang/dafny/issues/1607

> - feat: support variant type parameters on datatype definitions, which enables using traits as type arguments. Support for downcasting datatypes with contra-variant type parameters is left out. https://github.com/dafny-lang/dafny/issues/1499

Reviewer: ```suggestion
- feat: Support variant type parameters on datatype definitions, which enables using traits as type arguments. https://github.com/dafny-lang/dafny/issues/1499
- feat: Support for downcasting both custom datatypes and functions.
```

# 3.4 

The top four improvements in Dafny 3.4 are:

> - For certain classes of changes, prevent unexpected changes in verification behavior when changing a Dafny program.

Reviewer 2: ```suggestion
- For certain classes of changes to a Dafny program, prevent unexpected changes in verification behavior.
```

Just to avoid the slight redundancy.

- Add command line options to assist in debugging verification performance.
- Critical fixes to the IDE and greatly improved responsiveness of non-verification IDE features. 
- The C# back-end supports traits as type parameters on datatypes.

### Verification
- fix: Resolve unsoundness issue (https://github.com/dafny-lang/dafny/issues/1619).
- fix: Don't silently succeed if the solver crashes (https://github.com/boogie-org/boogie/pull/488).
- feat: Prevent changes in the verification behavior of a proof, when any of these types of changes are made to Dafny user code:

>   - Changes logically unrelated to the proof being verified

Reviewer 2: ```suggestion
  - Changes to declarations not referenced by the method being verified
```

The original claim is too strong, since it's not yet true for unrelated changes within the same method - @keyboardDrummer or @RustanLeino please confirm if I've over-simplified in my version.

Reviewer 3: Yeah, I like the updated wording, too.

  - Changes to the name of any declaration
  - Changes to the order of top-level declarations

> - feat: Assist in debugging the verification performance of a proof by adding the `/vcsSplitOnEveryAssert` CLI option and `{:vcs_split_on_every_assert}` attribute, and report the outcome and duration of splits when they occur in `/verificationLogger:trx` content.

Reviewer 2: ```suggestion
- feat: Assist in debugging the verification performance of a proof by adding the `/vcsSplitOnEveryAssert` CLI option and `{:vcs_split_on_every_assert}` attribute (see https://github.com/boogie-org/boogie/issues/465), and report the outcome and duration of splits when they occur in `/verificationLogger:trx` content.
```

Seems fair to note that the first part is actually a Boogie feature (especially since we'll want to make it a much more polished Dafny feature soon: https://github.com/dafny-lang/dafny/issues/1754)

- feat: Add a `/verificationLogger:csv` CLI option that emits the same status and timing information as `/verificationLogger:trx`, but in an easier-to-parse format, along with Z3 resource counts for more repeatable tracking of verification difficulty.

### IDE 
- fix: Clean up process resources if IDE closed or restarted.
- fix: Do not let the Dafny compilation status bar get in a stuck state.

> 

Reviewer 2: Minor: I slightly prefer listing the features before the bug fixes - it's a little more optimistic :)

- feat: Verification status reporting shows which proof is being verified, which can help debug slow to verify proofs.
- feat: Publish parsing and resolution diagnostics before verification has completed. Verification diagnostics from previous runs are migrated.
- feat: Enable 'go to definition', 'hover' and 'signature help' features before verification has completed.
- feat: Improve the hover feature to work for a wider scope of Dafny constructs, including function and method parameters, forall, exists and let expressions, and set and map comprehensions.
- feat: Add an experimental verification caching feature, which enables automatically determining which proofs need to verify again after making changes.
- feat: Display related resolution errors using nested diagnostics instead of independent diagnostics.

### UX

> - feat: Improve error reporting when providing incorrectly typed arguments in a function call.

Reviewer 2: ```suggestion
- feat: Improve error reporting when providing incorrectly-typed arguments in a function call.
```

> - feat: Improve error reporting when providing incorrectly-typed arguments in a function call.

Reviewer 3: ```suggestion
- feat: Improve error reporting when providing incorrectly typed arguments in a function call.
```

- feat: Improve error reporting when using type tests.

### C#

> - fix: resolve various instances where Dafny would produce invalid C# code (#1607, #1761, and #1762).

Reviewer 2: ```suggestion
- fix: Resolve various instances where Dafny would produce invalid C# code (https://github.com/dafny-lang/dafny/issues/1607, https://github.com/dafny-lang/dafny/issues/1761, and https://github.com/dafny-lang/dafny/issues/1762).
```

- feat: Support variant type parameters on datatype definitions, which enables using traits as type arguments (https://github.com/dafny-lang/dafny/issues/1499).

> - feat: Support for downcasting both custom datatypes and functions.

Reviewer 2: Could we be more specific and/or link the issue here? The description is a little ambiguous - does it mean explicit `is` tests, or implicit downcasting because the verifier proved no `is` is necessary?

Reviewer 5: Probably for @fabiomadge to answer?

Reviewer: This should affect both explicit `as` casts and implicit ones. A runtime check using `is` is unrelated.

### Various improvements
- fix: `DafnyLanguageServer.dll` and `Dafny.dll` depended on two different versions of Newtonsoft.Json, which could cause crashes in development environments.
- fix: (error reporting) Types with the same name but from different modules are now disambiguated in error messages.
- fix: (error reporting) Messages about arguments / parameters type mismatch are clearer and include the parameter name if available.
- fix: (robustness) Exceptions during parsing, if any, won't crash the language server anymore.
- fix: The elephant operator (`:-`) has a clearer error message and no longer reject generic methods on its right-hand side.

## Breaking changes

> - Proofs such as methods and lemmas whose verification behavior, either the result or the verification time, depend on arbitrary behavior in Dafny's solver, may show different verification behavior in Dafny 3.4. For users who are not ready to change those proofs to make their verification more reliable, we offer the command line option `/mimicVerificationOf:3.3` to keep the Dafny 3.4 verification behavior consistent with 3.3.

Reviewer 2: ```suggestion
- The verifier in Dafny 3.4 is now more efficient for many programs, and making changes to Dafny programs is less likely to cause verification to take longer or timeout. However, it is still possible for some correct programs to take longer to verify than on Dafny 3.3, or for verification to fail. For users with such programs who are not yet ready to modify them to pass the 3.4 verifier, we offer the command line option `/mimicVerificationOf:3.3` to keep the Dafny 3.4 verification behavior consistent with 3.3.
```

Saying the verification result can depend on arbitrary behavior in Dafny's solver seems to imply that invalid code might successfully verify.
